core/src/main/java/io/grpc/util/RoundRobinLoadBalancerFactory.java

```diff
@@ -211,21 +236,30 @@ public void shutdown() {
      */
     private void updateBalancingState(ConnectivityState state, Status error) {
       List<Subchannel> activeList = filterNonFailingSubchannels(getSubchannels());
-      helper.updateBalancingState(state, new Picker(activeList, error, stickinessState));
+      ready = !activeList.isEmpty();
+      if (state == null) {
+        state = ready ? READY : getAggregatedState();
+      }
+      SubchannelPicker picker = ready ? new Picker(activeList, stickinessState) :
+          new EmptyPicker(error != null ? error : getAggregatedError());
```

Comment:
there is a lot of conditional inline logic here which is hard to visually follow.  For example, why does `error` take precedence over aggregated error?  If the params are meant to be optional, it would be better expressed as an overload.

Reply:
Agree, I should restructure this to be more explicit/readable.

Some of it is to avoid unnecessary computation - for example `getAggregatedState()` is only applicable when the filtered `activeList` is empty, and not in the `handleNameResolutionError()` callback case. Similar thing for `error` -  it only applies in the empty active list case, where it's determine by `getAggregatedError()` _unless_ this is triggered by `handleNameResolutionError()` in which case the name resolver error is relayed.